### PR TITLE
Update tutorial examples to thread explicit RNGs

### LINKED ISSUE #4971: Emphasize Explicit Randomness in Documentation
The implicit RNG state used in Flax's documentation at the moment often ends up tripping up users when it comes to threading that state through transformations. We should steer users towards using explicit RNG states in their calls to dropout and other stochastic layers so that transformations become easier. This requires modifying the "Randomness", "Flax Basics", and "MNIST Tutorial" guides.

### PR DESCRIPTION
**Pull request by samanklesaria**
Closes #4971

### CODE REVIEW

**Comment by samanklesaria** [comment]
When using Jupyter to edit the notebooks, I find that all the `{code-cell}`  blocks in the markdown get replaced with `{code-cell} ipython3`. That is, jupytext on my machine labels all code cells with the language, while on the main branch, code cells are inconsistently labeled. In the mnist tutorial there's no `ipython3` label, but in the randomness tutorial there is. Instead of figuring out how to configure my jupyter settings separately for each notebook file, I'm just going to leave this commit including the extra `ipython3` labels.

File: docs_nnx/guides/randomness.md [hunk 1]
```diff
@@ -134,9 +138,80 @@ z1 = rngs.normal((2, 3))                 # generates key from rngs.default
...
+    return nnx.relu(self.drop(self.linear(x), rngs=rngs))
+```
+
+```{code-cell} ipython3
+model =  Model(rngs=nnx.Rngs(0))
+```
+
+```{code-cell} ipython3
+@nnx.vmap(in_axes=(None, 0, 0), out_axes=0)
+def train_step(model, x, rngs):
```
Comment (cgarciae) [inline docs_nnx/guides/randomness.md]: maybe use a `forward` function as a more realistic example and actually return the output?

File: docs_nnx/guides/randomness.md [hunk 2]
```diff
@@ -134,9 +138,80 @@ z1 = rngs.normal((2, 3))                 # generates key from rngs.default
...
+dropout_rngs
+```
+
+```{code-cell} ipython3
+forked_rngs = dropout_rngs.fork(split=5)
+```
+
+```{code-cell} ipython3
+forked_rngs
+```
```
Comment (cgarciae) [inline docs_nnx/guides/randomness.md]: merge with previous cell? (also `dropout_rngs` cell)

File: docs_nnx/guides/randomness.md [hunk 3]
```diff
@@ -134,9 +138,80 @@ z1 = rngs.normal((2, 3))                 # generates key from rngs.default
...
+
+  def __call__(self, x):
+    return nnx.relu(self.drop(self.linear(x)))
+
+model = Model(nnx.Rngs(params=0, dropout=1))
+
+y = model(x=jnp.ones((1, 20)))
+print(f'{y.shape = }')
+```
+
```
Comment (cgarciae) [inline docs_nnx/guides/randomness.md:212]: It would be great to discuss the pros and cons of implicit state after showing the example.

File: docs_nnx/guides/randomness.md [hunk 4]
```diff
@@ -180,12 +255,7 @@ assert jnp.allclose(y1, y3)     # same
 
 ## Splitting PRNG keys
 
-When interacting with [Flax NNX transforms](https://flax.readthedocs.io/en/latest/guides/transforms.html) like `nnx.vmap` or `nnx.pmap`, it is often necessary to split the random state such that each replica has its own unique state. This can be done in two ways:
-
-- By manually splitting a key before passing it to one of the `nnx.Rngs` streams; or
-- By using the `nnx.split_rngs` decorator which will automatically split the random state of any `nnx.RngStream`s found in the inputs of the function, and automatically "lower" them once the function call ends.
-
-It is more convenient to use `nnx.split_rngs`, since it works nicely with Flax NNX transforms, so here’s one example:
+When interacting with [Flax NNX transforms](https://flax.readthedocs.io/en/latest/guides/transforms.html) like `nnx.vmap` or `nnx.pmap`, it is often necessary to split the implicit random state such that each replica has its own unique state. This can be done using the `nnx.split_rngs` decorator which will automatically split the random state of any `nnx.RngStream`s found in the inputs of the function, and automatically "lower" them once the function call ends. Here’s an example:
```
Comment (cgarciae) [inline docs_nnx/guides/randomness.md]: Lets replace `nnx.split_rngs` in this example bellow with the newer `nnx.fork_rngs`. Easier to understant now that we've explained `Rngs.fork` above.

**Comment by cgarciae** [comment]
Thanks @samanklesaria ! Its looking great. Left some comments.
Maybe the other place we can update is the `index.rst` which also contains an example that uses dropout.

**Comment by cgarciae** [comment]
Thanks @samanklesaria ! Wondering if we could also refactor the intro to the Randomness guide as part of this PR? Current intro is a bit outdated in the sense that it tries to explain it in terms of Haiku and Linen. It might be more useful add a simple code example at the beginning as motivation for the whole guide e.g:

```python
class Model(nnx.Module):
  def __init__(self, *, rngs: nnx.Rngs):
    self.linear = nnx.Linear(20, 10, rngs=rngs)
    self.drop = nnx.Dropout(0.1)

  def __call__(self, x, *, rngs):
    return nnx.relu(self.drop(self.linear(x), rngs=rngs))
 
rngs = nnx.Rngs(0)
model = Model(rngs=rngs)  # pass rngs to initialize parameters
x = rngs.normal((32, 20))  # convenient jax.random methods
y = model(x, rngs=rngs)  # pass rngs for dropout masks
```

We can remove the current section below in favor of the example above and some comments introducing `nnx.Rngs` as the main mechanism that NNX uses for propagating random state.

```
Random state handling in Flax NNX was radically simplified compared to systems like Haiku and Flax Linen because Flax NNX defines the random state as an object state. In essence, this means that in Flax NNX, the random state is: 1) just another type of state; 2) stored in nnx.Variables; and 3) held by the models themselves.

The Flax NNX pseudorandom number generator (PRNG) system has the following main characteristics:

It is explicit.

It is order-based.

It uses dynamic counters.

This is a bit different from Flax Linen’s PRNG system, which is (path + order)-based, and uses static counters.
```

**Comment by Tsukimarf** [comment]
ns module methods under jax.named_scope for profiling
